src/libcore/iter/iterator.rs

```diff
@@ -2523,6 +2523,95 @@ pub trait Iterator {
             }
         }
     }
+
+    /// Checks if the elements of this iterator are sorted.
+    ///
+    /// That is, for each element `a` and its following element `b`, `a <= b` must hold. If the
+    /// iterator yields exactly zero or one element, `true` is returned.
+    ///
+    /// Note that if `Self::Item` is only `PartialOrd`, but not `Ord`, the above definition
+    /// implies that this function returns `false` if any two consecutive items are not
+    /// comparable.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(is_sorted)]
+    ///
+    /// assert!([1, 2, 2, 9].iter().is_sorted());
+    /// assert!(![1, 3, 2, 4].iter().is_sorted());
+    /// assert!([0].iter().is_sorted());
+    /// assert!(std::iter::empty::<i32>().is_sorted());
+    /// assert!(![0.0, 1.0, std::f32::NAN].iter().is_sorted());
+    /// ```
+    #[inline]
+    #[unstable(feature = "is_sorted", reason = "new API", issue = "53485")]
+    fn is_sorted(self) -> bool
+    where
+        Self: Sized,
+        Self::Item: PartialOrd,
+    {
+        self.is_sorted_by(|a, b| a.partial_cmp(b))
+    }
+
+    /// Checks if the elements of this iterator are sorted using the given comparator function.
+    ///
+    /// Instead of using `PartialOrd::partial_cmp`, this function uses the given `compare`
+    /// function to determine the ordering of two elements. Apart from that, it's equivalent to
+    /// [`is_sorted`]; see its documentation for more information.
+    ///
+    /// [`is_sorted`]: trait.Iterator.html#method.is_sorted
+    #[unstable(feature = "is_sorted", reason = "new API", issue = "53485")]
+    fn is_sorted_by<F>(mut self, mut compare: F) -> bool
```

Comment:
Does the same reasoning apply for `[T]`'s implementation of `is_sorted_by`? I can change both back, but I also originally had both as `#[inline]` so I just want to be sure before making the change again.

Reply:
Yeh, I think this one was my bad, I steered you off course. I hadn't realised that when an `#[inline]`'ed function internally calls a generic, but not explicitly `#[inline]`'ed, function then that inner function won't be inlined at all.

Reply:
Arguably, the case in which this won't be inlined is contrived, and these are currently not used anywhere inside `core` yet, so that case cannot happen until then. Also, I don't think the auto-vectorizer can currently handle these methods due to the early return, so whether these are inlined or not might not be that big of a deal. 

However, if someone decides to use these somewhere inside core, chances are that they won't think about any of this, and `#[inline]` doesn't really make anything any worse since they are already generic, so I'd rather just make them `#[inline]` now.